[EMAIL-207] Modify createDataSource method to use AttachmentDataSource

violetbeach opened the pull request:
### Issue

Hello,

OOM occurred in the mail service developed using Apache-commons-email.

I debugged and found that the mail lookup function was wasting memory!

### Problem

Below is a test for reproducing the problem.

```java
@Test
void memoryTest() throws Exception {
    InputStream is = new SharedFileInputStream("/test/oom.eml");
    Session session = Session.getInstance(System.getProperties());
    MimeMessage mimeMessage = new MimeMessage(session, is);
    MimeMessageParser parser = new MimeMessageParser(mimeMessage).parse();
}
```

Below is the memory usage.

![img](https://github.com/apache/commons-email/assets/63458653/721ddcb7-b30b-4688-afc4-6f773c85b6a8)

I expected to write attachments to the buffer of the output stream when absolutely necessary.

However, it is becoming a problem by putting all bytes[] in memory.

![img_3](https://github.com/apache/commons-email/assets/63458653/67c7f208-fc4a-4954-823a-3a6904d21dd7)

![img_1](https://github.com/apache/commons-email/assets/63458653/8b144f5a-2b15-4c9a-a070-4d179e0b1d09)

### Sugest

So i modified these processes to use AttachmentDataSource.

Attachment DataSource uses InputStream from original DataSource to ensure memory is not wasted.

As a result, i can get a big effect in memory usage.

![img_2](https://github.com/apache/commons-email/assets/63458653/9adb56e3-8a57-49ad-aec7-1324860b0f30)

> I worked hard to fix it. Feel free to let me know if you have any advice or lack!

If you need the tested eml file, I will send it to you by email.

Thanks very much. 🙇

garydgregory reviewed (changes requested):
@violetbeach 
Thank you for your PR.
- All new public and protected elements should have a Javadoc comment
- All new and changed code should have a matching unit test

violetbeach commented:
> * All new public and protected elements should have a Javadoc comment
> * All new and changed code should have a matching unit test

Thank very much!

I added Javadoc comments on public and protected elements. And I added a unit test for new class.

But, i could not cover the changed createDataSource method of MimeMessageParser because it is a difficult area to test.
I could only check that all the existing tests worked well.

Please let me know if there's anything missing! Thank you.

garydgregory commented on src/main/java/org/apache/commons/mail/util/AttachmentDataSource.java:
```diff
@@ -0,0 +1,104 @@
+/*
```
Looks like a copy of https://github.com/apache/cxf/blob/main/rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/ext/multipart/InputStreamDataSource.java ?

violetbeach replied:
Oh! It's almost the same.

garydgregory commented:
Hello @violetbeach 
This issue has been handled slightly differently in git master. You are credited in `changes.xml`.

violetbeach commented:
Hello @garydgregory 

I've verified that the issue has been solved!

<img width="1776" alt="image" src="https://github.com/apache/commons-email/assets/63458653/53c57fef-caf3-48a9-86dc-d586a3a46f90">

Memory is much cleaner, and the parser no longer puts the entire data in the file into memory.

Thank you. 👍

garydgregory commented:
@violetbeach 👍
